src/localconstraints/local_forbidden_sequence.jl

```diff
@@ -0,0 +1,145 @@
+    #TODO: it would be better to precompute the list of `nodes` on the path.
+    # however, in the GenericSolver, uniform holes might be replaced with a rule node, so the instances might change
+    nodes = get_nodes_on_path(get_tree(solver), c.path)
+    track!(solver, "LocalForbiddenSequence propagation")
+
+    # c.sequence = [1, 2, 3]
+    # c.nodes = [RuleNode(1), RuleNode(2), UniformHole(1, 2), UniformHole(2, 4), RuleNode(3)]
+    # c.nodes = [1, {2, 4}, 1, {2, 4}, 3]
+
+    # Smallest match
+    forbidden_assignments = Vector{Tuple{Int, Any}}()
+    i = length(c.sequence)
+    for (path_idx, node) ∈ Iterators.reverse(enumerate(nodes))
+        forbidden_rule = c.sequence[i]
+        if (node isa RuleNode) || (node isa StateHole && isfilled(node)) #TODO: in herbcore: isfilled(::AbstractHole) = false
+            rule = get_rule(node)
+            if (rule ∈ c.ignore_if)
+                deactivate!(solver, c)
+                track!(solver, "LocalForbiddenSequence deactivate by ignore_if")
+                return
+            elseif (rule == forbidden_rule)
+                i -= 1
+            end
+        else
+            if node.domain[forbidden_rule]
+                push!(forbidden_assignments, (path_idx, forbidden_rule))
+                i -= 1
+            else
+                for r ∈ c.ignore_if
+                    if node.domain[r]
+                        rules = [r for r ∈ findall(node.domain) if r ∉ c.ignore_if]
+                        if !isempty(rules)
+                            push!(forbidden_assignments, (path_idx, rules))
+                            i -= 1
+                        end
+                        break
+                    end
+                end
+            end
+        end
+        if i == 0
+            break
+        end
+    end
+    if i > 0
+        track!(solver, "LocalForbiddenSequence deactivate")
+        deactivate!(solver, c)
+        return
+    end
+    if length(forbidden_assignments) == 0
+        track!(solver, "LocalForbiddenSequence inconsistency")
+        set_infeasible!(solver)
+        return
+    elseif length(forbidden_assignments) == 1
+        path_idx, rule = forbidden_assignments[1]
+        if rule isa Int
+            track!(solver, "LocalForbiddenSequence deduction")
+        else
+            track!(solver, "LocalForbiddenSequence deduction by ignore_if")
+        end
+        if path_idx > length(c.path)
+            deactivate!(solver, c)
+        end
+        remove!(solver, c.path[1:path_idx-1], rule)
+        return
+    end
+
+    # Smallest match with a maximum of 1 hole (Optional, slightly stronger inference without making all possible matches) 
+    i = length(c.sequence)
+    forbidden_assignment = nothing
+    for (path_idx, node) ∈ Iterators.reverse(enumerate(nodes))
+        forbidden_rule = c.sequence[i]
+        if (node isa RuleNode) || (node isa StateHole && isfilled(node)) #TODO: in herbcore: isfilled(::AbstractHole) = false
```

Comment:
Issue in `HerbCore` for this too, please

Reply:
This is now an issue in `HerbCore`. I deleted the todo.
The reason I added it here, was to remind myself of a hacky workaround.

This line should be:
```julia
if isfilled(node)
```

But since isfilled(node) for non-state holes is $O(n)$, I skip that check because I can assume that non-state holes have a domain size of >= 2 (otherwise they are replaced with a rule node by the generic solver). So I use a hacky check:

```julia
if (node isa RuleNode) || (node isa StateHole && isfilled(node))
```

Case (A): node isa RuleNode, so isfilled(node) is true.
Case (B): node isa StateHole , so we check isfilled(node).
Case (C): node isa Hole, I assume this is not filled, otherwise it was replaced with a RuleNode. so isfilled(node) is false.
Case (D): node isa UniformHole. Same as (C).

Reply:
Got it. Just gonna link the issue here in case we come across this comment in the future: https://github.com/Herb-AI/HerbCore.jl/issues/22